pageserver/src/tenant/storage_layer/delta_layer.rs

```diff
@@ -847,84 +849,102 @@ impl DeltaLayerInner {
+            .build();
 
         for range in keyspace.ranges.iter() {
             let mut range_end_handled = false;
 
             let start_key = DeltaKey::from_key_lsn(&range.start, lsn_range.start);
-            tree_reader
-                .visit(
-                    &start_key.0,
-                    VisitDirection::Forwards,
-                    |raw_key, value| {
-                        let key = Key::from_slice(&raw_key[..KEY_SIZE]);
-                        let lsn = DeltaKey::extract_lsn_from_buf(raw_key);
-                        let blob_ref = BlobRef(value);
-
-                        assert!(key >= range.start && lsn >= lsn_range.start);
-
-                        let cached_lsn = reconstruct_state.get_cached_lsn(&key);
-                        let flag = {
-                            if cached_lsn >= Some(lsn) {
-                                BlobFlag::Ignore
-                            } else if blob_ref.will_init() {
-                                BlobFlag::Replaces
-                            } else {
-                                BlobFlag::None
-                            }
-                        };
-
-                        if key >= range.end || (key.next() == range.end && lsn >= lsn_range.end) {
-                            planner.handle_range_end(blob_ref.pos());
-                            range_end_handled = true;
-                            false
-                        } else {
-                            planner.handle(key, lsn, blob_ref.pos(), flag);
-                            true
-                        }
-                    },
-                    &RequestContextBuilder::extend(ctx)
-                        .page_content_kind(PageContentKind::DeltaLayerBtreeNode)
-                        .build(),
-                )
-                .await
-                .map_err(|err| anyhow!(err))?;
+            let index_stream = index_reader.get_stream_from(&start_key.0, &btree_request_context);
+            pin_mut!(index_stream);
+
+            while let Some(index_entry) = index_stream.next().await {
+                let (raw_key, value) = index_entry.map_err(|err| anyhow!(err))?;
+                let key = Key::from_slice(&raw_key[..KEY_SIZE]);
+                let lsn = DeltaKey::extract_lsn_from_buf(&raw_key);
+                let blob_ref = BlobRef(value);
+
+                // Lsns are not monotonically increasing, so we don't assert on them.
+                assert!(key >= range.start);
+
+                let flag = {
+                    #[allow(clippy::if_same_then_else)]
+                    if lsn >= lsn_range.end || lsn < lsn_range.start {
+                        // If the Lsn is not in the queried range it must be ignored
+                        BlobFlag::Ignore
+                    } else if reconstruct_state.get_cached_lsn(&key) >= Some(lsn) {
+                        // If the Lsn is below the caching line it must be ignored
+                        BlobFlag::Ignore
```

Comment:
> Does the "below caching line" mean that we've already found an older walrecord for this key?

No. The non-vectored read path will first check the materialised page cache to see if the requested key is already cached at some Lsn. This can make the search path shorter. The vectored read path doesn't use the materialized page cache, but it allows for the same functionality such that we'll be able to switch over to the vectored impl everywhere and still use the materialized page cache.

Reply:
Thanks, that makes sense. Resolving.